analysis.ml

```diff
@@ -177,26 +177,27 @@ let reaching (instrs : segment) : pc -> PcSet.t =
     let all_definitions = List.map definitions_of used in
     List.fold_left PcSet.union PcSet.empty all_definitions
 
-
-let liveness_analysis (instrs : segment) =
-  let merge _pc cur_uses in_uses =
-    let merged = VariableMap.union cur_uses in_uses in
-    if VariableMap.equal cur_uses merged then None else Some merged
+let scope_analysis (introduction : instruction -> variable list)
+                   (elimination : instruction -> variable list) (instrs : segment) =
+  let merge _pc cur_scope in_scope =
+    let merged = VariableMap.union cur_scope in_scope in
+    if VariableMap.equal cur_scope merged then None else Some merged
   in
-  let update pc uses =
+  let update pc cur_scope =
     let instr = instrs.(pc) in
-    (* First remove defined vars *)
-    let kill = VarSet.elements (ModedVarSet.untyped (defined_vars instr)) in
+    let to_remove = introduction instr in
     let remove acc var = VariableMap.remove var acc in
-    let uses = List.fold_left remove uses kill in
-    (* Then add used vars *)
-    let used = VarSet.elements (used_vars instr) in
-    let merge acc var = VariableMap.union (VariableMap.singleton var pc) acc
-    in
-    List.fold_left merge uses used
+    let cur_scope' = List.fold_left remove cur_scope to_remove in
+    let to_add = elimination instr in
+    let merge acc var = VariableMap.union (VariableMap.singleton var pc) acc in
+    List.fold_left merge cur_scope' to_add
```

Comment:
`List.fold_right (fun v -> VariableMap.add v pc) to_add cur_scope` works here (it's a bit clever, but idiomatic)